tools/qlog/src/streamer.rs

```diff
@@ -74,56 +80,105 @@ impl QlogStreamer {
+                        .to_format_label(),
+                    title,
+                    description,
+                    summary,
+                    trace,
+                };
+
+                QlogStreamer {
+                    start_time,
+                    writer,
+                    qlog: None,
+                    qlog_seq: Some(qlog_seq),
+                    state: StreamerState::Initial,
+                    log_level,
+                    first_event: true,
+                    first_frame: false,
+                }
+            },
         }
     }
 
     /// Starts qlog streaming serialization.
     ///
-    /// This writes out the JSON-serialized form of all information up to qlog
-    /// `Trace`'s array of `Event`s. These are are separately appended
-    /// using `add_event()` and `add_event_with_instant()`.
+    /// This writes out the JSON-serialized form of all initial qlog information
+    /// `Event`s are separately appended using `add_event()` and
+    /// `add_event_with_instant()`.
     pub fn start_log(&mut self) -> Result<()> {
         if self.state != StreamerState::Initial {
             return Err(Error::Done);
         }
 
-        // A qlog contains a trace holding a vector of events that we want to
-        // serialize in a streaming manner. So at the start of serialization,
-        // take off all closing delimiters, and leave us in a state to accept
-        // new events.
-        match serde_json::to_string(&self.qlog) {
-            Ok(mut out) => {
-                out.truncate(out.len() - 4);
+        if let Some(ref qlog) = self.qlog {
+            // A `Qlog` contains a `Trace` holding a vector of events that we want
+            // to serialize in a streaming manner. So at the start of
+            // serialization, take off all closing delimiters, and
+            // leave us in a state to accept new events.
+            match serde_json::to_string(qlog) {
+                Ok(mut out) => {
+                    out.truncate(out.len() - 4);
 
-                self.writer.as_mut().write_all(out.as_bytes())?;
+                    self.writer.as_mut().write_all(out.as_bytes())?;
 
-                self.state = StreamerState::Ready;
+                    self.state = StreamerState::Ready;
 
-                self.first_event = self.qlog.traces[0].events.is_empty();
-            },
+                    if let Some(ref qlog) = self.qlog {
+                        self.first_event = qlog.traces[0].events.is_empty();
+                    }
+                },
 
-            _ => return Err(Error::Done),
+                _ => return Err(Error::Done),
+            }
+        }
+
+        if let Some(ref qlog_seq) = self.qlog_seq {
+            // The `QlogSeq` contains a simple `TraceSeq`, so we can write
+            // it out directly as a standalone item.
+
+            match serde_json::to_string(qlog_seq) {
+                Ok(out) => {
+                    let out = format!("{}\n", out);
```

Comment:
In alternative I also just realized it would probably be even better to replace all the `serde_json::to_string()` calls with `serde_json::to_writer()`, which would avoid additional temporary allocations and make things more efficient https://docs.rs/serde_json/1.0.69/serde_json/fn.to_writer.html

It can probably be left for a future refactoring though, so we can either do the `write!()` change I suggested above now, and `to_writer()` later, or just do `to_writer()` now, but that's probably slightly trickier as it changes the code structure a bit.

Reply:
This all becomes easier if we just write out qlog events atomically. Building on top of this branch, here is another branch that would change to atomic events that are written with `serde_json::to_writer()` - https://github.com/cloudflare/quiche/compare/lucas/json-seq...lucas/atomic-json-events?expand=1